crates/maple_cli/src/cmd/filter.rs

```diff
@@ -0,0 +1,499 @@
+                    })
+                }),
+                number,
+            ),
+            Source::File(fpath) => dyn_collect_number(
+                std::fs::read_to_string(fpath)?.lines().filter_map(|line| {
+                    scorer(&line).map(|(score, indices)| (line.into(), score, indices))
+                }),
+                number,
+            ),
+            Source::List(list) => dyn_collect_number(
+                list.filter_map(|line| {
+                    scorer(&line).map(|(score, indices)| (line, score, indices))
+                }),
+                number,
+            ),
+        };
+
+        let ranked = filtered;
+
+        let payload = ranked.into_iter().take(number);
+        let winwidth = winwidth.unwrap_or(62);
+        let (truncated_payload, truncated_map) =
+            truncate_long_matched_lines(payload, winwidth, None);
+        let mut lines = Vec::with_capacity(number);
+        let mut indices = Vec::with_capacity(number);
+        if enable_icon {
+            for (text, _, idxs) in truncated_payload {
+                let iconized = prepend_icon(&text);
+                lines.push(iconized);
+                indices.push(idxs);
+            }
+        } else {
+            for (text, _, idxs) in truncated_payload {
+                lines.push(text);
+                indices.push(idxs);
+            }
+        }
+        if truncated_map.is_empty() {
+            println_json!(total, lines, indices);
+        } else {
+            println_json!(total, lines, indices, truncated_map);
+        }
+    } else {
+        let mut filtered = match source {
+            Source::Stdin => dyn_collect_all(io::stdin().lock().lines().filter_map(|lines_iter| {
+                lines_iter
+                    .ok()
+                    .and_then(|line| scorer(&line).map(|(score, indices)| (line, score, indices)))
+            })),
+            Source::File(fpath) => {
+                dyn_collect_all(std::fs::read_to_string(fpath)?.lines().filter_map(|line| {
+                    scorer(line).map(|(score, indices)| (line.into(), score, indices))
+                }))
+            }
+            Source::List(list) => {
+                dyn_collect_all(list.filter_map(|line| {
+                    scorer(&line).map(|(score, indices)| (line, score, indices))
+                }))
+            }
+        };
+
+        filtered.par_sort_unstable_by(|(_, v1, _), (_, v2, _)| v2.partial_cmp(&v1).unwrap());
+
+        let ranked = filtered;
+
+        for (text, _, indices) in ranked.iter() {
+            println_json!(text, indices);
+        }
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    // This is a very time-consuming test,
+    // results of which could be proved only be inspecting stdout.
+    // Definetly not something you want to run with `cargo test`.
+    #[ignore]
+    fn dynamic_results() {
```

Comment:
Have not got time to look at the test as I have to merge the latest mater/rebase this PR as it has too much code that are unrelated. Will let you know when I tried the test.

Reply:
> 
> 
> Have not got time to look at the test as I have to merge the latest mater/rebase this PR as it has too much code that is unrelated. Will let you know when I try the test.

It's just `git clone https://github.com/ImmemorConsultrixContrarie/vim-clap` and `cd vim-clap/crates` with `cargo test -- --ignored --nocapture`.

Reply:
@ImmemorConsultrixContrarie  Just grab the code and run the test, it just works the way that I have always wanted for dynamic filter, awesome job! :+1: 

I suggest using `Source::File(random_generated_data_file)` for the test purpose as we could compare the final result of dynamic version and the non-dynamic version and that's one of the real use case, and it's more readable personally. But it's ok if you're too lazy for this, I can take the test job anyway. For me, I use this locally:

```bash
$ cd / && fd --type f > ~/root.txt
$ # Use Source::File(`~/root.txt`) in the test
```

Once we get this big feature merged, we can also expand the dynamic feature further. The ultimate imagtination is the `Source` can be generated in async, like we spawn the command `fd --type f` in a thread and use its stdout as the source of our filter, do the filtering in another thread.

Reply:
> I suggest using `Source::File(random_generated_data_file)` for the test purpose as we could compare the final result of dynamic version and the non-dynamic version and that's one of the real use case

Yeah, but this should be really big file and I don't have any. And you can just make this generator determined, if you write `let bytes = 123;` instead of time-dependent bytes.

Reply:
Okay, if it works, I guess this draft should be finally merged with master branch, so next commits won't be that painful to work with, so I'm gonna mark this draft as finally ready for review.

Reply:
@ImmemorConsultrixContrarie Actually it's easy to get the testdata it I have used it before :(.

```bash
$ vim-clap/test/bench/python
$ bash fetch_testdata.sh
$ # Now you'll see testdata.txt
```

This test file has about 850K lines and has beenn ignored in .gitignore.

Reply:
@ImmemorConsultrixContrarie Since you have marked this PR reviewable, can you please resolve the conflicts also? Except the icon related code, most of them should just take yours.